src/lib.rs

```diff
@@ -0,0 +1,386 @@
+            Channel::Pro(ch) => ch.handle(),
+        }
+    }
+
+    fn info(&self) -> &ChannelInfo {
+        match self {
+            Channel::Eco(ch) => ch.info(),
+            Channel::Pro(ch) => ch.info(),
+        }
+    }
+}
+
+pub struct EcoChannel {
+    handle: *mut c_void,
+    info: ChannelInfo,
+    event_fd: EventFd,
+    // hold onto a reference of the currently set capture frame
+    video_capture_frame: Option<Pin<Box<EcoVideoCaptureFrame>>>,
+}
+
+impl UniversalCaptureFamily for EcoChannel {
+    fn handle(&self) -> *mut c_void {
+        self.handle
+    }
+
+    fn info(&self) -> &ChannelInfo {
+        &self.info
+    }
+}
+
+impl ProEcoCaptureFamily for EcoChannel {}
+
+impl EcoChannel {
+    pub fn start_video_capture(&mut self, width: u16, height: u16, format: FourCC) -> Result<()> {
+        let mut params = sys::_MWCAP_VIDEO_ECO_CAPTURE_OPEN {
+            cx: width as _,
+            cy: height as _,
+            dwFOURCC: format.as_u32(),
+            llFrameDuration: -1,
+            hEvent: self.event_fd.as_raw_fd() as _,
+        };
+        unsafe {
+            if sys::MWStartVideoEcoCapture(self.handle(), &mut params as *mut _)
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to start video capture");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn stop_video_capture(&mut self) -> Result<()> {
+        unsafe {
+            if sys::MWStopVideoEcoCapture(self.handle()) != sys::_MW_RESULT__MW_SUCCEEDED {
+                whatever!("unable to stop video capture");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn set_video_capture_frame(&mut self, frame: EcoVideoCaptureFrame) -> Result<()> {
+        let mut frame = Box::pin(frame);
+        unsafe {
+            if sys::MWCaptureSetVideoEcoFrame(self.handle(), frame.as_mut_ptr())
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to set video frame");
+            }
+        }
+        self.video_capture_frame = Some(frame);
+        Ok(())
+    }
+
+    pub fn get_video_capture_status(&mut self) -> Result<EcoVideoCaptureStatus> {
+        let status = unsafe {
+            let mut status = MaybeUninit::uninit();
+            if sys::MWGetVideoEcoCaptureStatus(self.handle(), status.as_mut_ptr())
+                != sys::_MW_RESULT__MW_SUCCEEDED
+            {
+                whatever!("unable to get video capture status");
+            }
+            status.assume_init()
+        };
+        Ok(EcoVideoCaptureStatus::new(
+            *Pin::into_inner(
+                self.video_capture_frame
+                    .take()
+                    .whatever_context("no video frame set")?,
+            ),
+            status,
+        ))
+    }
+
+    pub fn wait(&self) -> Result<()> {
+        if self
+            .event_fd
+            .read()
+            .whatever_context("unable to read eventfd")?
+            >= (1 << 32)
```

Comment:
Why this threshold? I skimmed docs and examples but it wasn't immediately obvious to me. (The examples seem to conflate the value returned here with the error paths which doesn't help.)

Reply:
I was trying to match the check in `EcoEventWait` in the examples, but on second look, I think this should be `== 0`. Changing.